### ISSUE TITLE
Investigate new Intel build failures

### ISSUE THREAD

**Issue by fcooper8472** [Chaste/Chaste#146]
E.g. https://github.com/Chaste/Chaste/actions/runs/5609769767/jobs/10263908241?pr=145

```
[ 11%] Building CXX object global/test/CMakeFiles/TestDebugParallel.dir/TestDebugParallel.cpp.o
In file included from /home/runner/work/Chaste/Chaste/chaste-build-dir/global/test/TestArchivingHelperClassesParallel.cpp:21:
/home/runner/work/Chaste/Chaste/global/test/TestArchivingHelperClasses.hpp:68:9: error: exception of type 'Exception &' will be caught by earlier handler [-Werror,-Wexceptions]
        TS_ASSERT_THROWS_THIS(ArchiveLocationInfo::GetArchiveDirectory(),
        ^
/home/runner/work/Chaste/Chaste/chaste-build-dir/cxxtest/cxxtest/TestSuite.h:429:39: note: expanded from macro 'TS_ASSERT_THROWS_THIS'
#   define TS_ASSERT_THROWS_THIS(e,y) CHASTE_ASSERT_EXCEPTION_MESSAGE(__FILE__,__LINE__,e,CheckShortMessage,y)
                                      ^
/home/runner/work/Chaste/Chaste/chaste-build-dir/cxxtest/cxxtest/TestSuite.h:423:56: note: expanded from macro 'CHASTE_ASSERT_EXCEPTION_MESSAGE'
            _TS_CATCH_TYPE( (const Exception& _err), { TSM_ASSERT(_err.method(y), _err.method(y) == ""); _ts_threw_expected = true; } ) \
                                                       ^
/home/runner/work/Chaste/Chaste/chaste-build-dir/cxxtest/cxxtest/TestSuite.h:[249](https://github.com/Chaste/Chaste/actions/runs/5609769767/jobs/10263908241?pr=145#step:8:250):28: note: expanded from macro 'TSM_ASSERT'
#   define TSM_ASSERT(m,e) _TSM_ASSERT(__FILE__,__LINE__,m,e)
                           ^
note: (skipping 2 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/runner/work/Chaste/Chaste/chaste-build-dir/cxxtest/cxxtest/TestSuite.h:209:32: note: expanded from macro '__TS_CATCH'
#       define __TS_CATCH(f,l) __TSM_CATCH(f,l,"Unhandled exception")
                               ^
/home/runner/work/Chaste/Chaste/chaste-build-dir/cxxtest/cxxtest/TestSuite.h:208:17: note: expanded from macro '__TSM_CATCH'
                ___TSM_CATCH(f,l,m)
                ^
/home/runner/work/Chaste/Chaste/chaste-build-dir/cxxtest/cxxtest/TestSuite.h:201:27: note: expanded from macro '___TSM_CATCH'
                    catch(Exception &_e) { (CxxTest::tracker()).failedTest(f,l,_e.GetMessage().c_str()); } \
                          ^
/home/runner/work/Chaste/Chaste/global/test/TestArchivingHelperClasses.hpp:68:9: note: for type 'const std::exception &'
/home/runner/work/Chaste/Chaste/chaste-build-dir/cxxtest/cxxtest/TestSuite.h:429:39: note: expanded from macro 'TS_ASSERT_THROWS_THIS'
#   define TS_ASSERT_THROWS_THIS(e,y) CHASTE_ASSERT_EXCEPTION_MESSAGE(__FILE__,__LINE__,e,CheckShortMessage,y)
...
```

Looks like some interaction between exceptions and our custom asserts. Possible it's related to 
https://github.com/Chaste/Chaste/pull/80 but that did have green ticks when merged...

**Comment by fcooper8472** [comment]
Possibly related to a new version of the intel compiler: the build failures are reported with 2023.2 whereas the previous green ticks were 2023.1.

**Comment by fcooper8472** [comment]
The problem is in `cxxtest/TestSuite.hpp`:

```cpp
#           define ___TSM_CATCH(f,l,m) \
                    catch(const std::exception &_e) { (CxxTest::tracker()).failedTest(f,l,_e.what()); } \
                    catch(Exception &_e) { (CxxTest::tracker()).failedTest(f,l,_e.GetMessage().c_str()); } \
                    _TSM_LAST_CATCH(f,l,m)
```

The error is actually quite informative, and appears to be correct:

```
 error: exception of type 'Exception &' will be caught by earlier handler [-Werror,-Wexceptions]
```

This appears to be a genuine error caught by the new compiler version, and relates to #80 : now that Chaste inherits from `std::exception`, I believe that second line is redundant.

Might be worth some input from @jmpf and @MILeach to check I'm not missing something dangerous.

**Comment by fcooper8472** [comment]
closed by #147
